cfme/tests/cloud/test_keypairs.py

```diff
@@ -73,41 +79,93 @@ def test_keypair_crud_with_key(provider, appliance):
-        * Provide Keypair name.
-        * Select Cloud Provider.
-        * Also delete it.
-
     Polarion:
-        assignee: mnadeem
-        casecomponent: WebUI
+        assignee: mmojzis
+        casecomponent: Cloud
+        caseimportance: low
         initialEstimate: 1/4h
+        testSteps:
+            1. Cancel creating keypair.
     """
     keypair = appliance.collections.cloud_keypairs.create(
         name="",
         provider=provider,
         cancel=True
     )
-
     assert not keypair.exists
 
 
+@pytest.mark.tier(3)
+def test_keypair_create_name_validation(provider, appliance):
+    """ This will test validating that key pair without name cannot be created.
+    Polarion:
+        assignee: mmojzis
+        casecomponent: Cloud
+        caseimportance: low
+        initialEstimate: 1/4h
+        testSteps:
+            1. Try to add key pair with empty name.
+    """
+    # test empty name
+    keypair_collection = appliance.collections.cloud_keypairs
+    view = navigate_to(keypair_collection, 'Add')
+    view.fill({'provider': provider})
+    assert not view.form.add.active
+
+    # test invalid name - on ec2 any name can be specified so only testing openstack
+    if provider.one_of(OpenStackProvider):
+        view = navigate_to(keypair_collection, 'Add')
+        keypair_name = '.?!_'
+        view.form.fill({'name': keypair_name,
+                   'provider': provider.name})
+        view.form.add.click()
+        view = view.browser.create_view(KeyPairAllView)
+        view.flash.assert_message('Unable to create Key Pair "{}": Keypair data is invalid: '
+                                  'Keypair name contains unsafe characters'.format(keypair_name))
+
+
+@pytest.mark.tier(3)
+def test_keypair_create_invalid_key_validation(provider, appliance):
+    """ This will test validating that key pair with invalid public key cannot be created.
+    Polarion:
+        assignee: mmojzis
+        casecomponent: Cloud
+        caseimportance: low
+        initialEstimate: 1/4h
+        testSteps:
+            1. Try to add key pair with invalid public key.
+    """
+    invalid_key = fauxfactory.gen_alphanumeric()
+    keypair_name = fauxfactory.gen_alphanumeric()
+    keypair_collection = appliance.collections.cloud_keypairs
+    view = navigate_to(keypair_collection, 'Add')
+    view.form.fill({'name': keypair_name,
+               'provider': provider.name,
+               'public_key': invalid_key})
+    view.form.add.click()
+    view = view.browser.create_view(KeyPairAllView)
+    if provider.one_of(EC2Provider):
+        view.flash.assert_message('Unable to create Key Pair "{}": Key is not in valid OpenSSH '
+                                  'public key format'.format(keypair_name))
+    elif provider.one_of(OpenStackProvider):
+        view.flash.assert_message('Unable to create Key Pair "{}": Keypair data is invalid: failed '
+                                  'to generate fingerprint'.format(keypair_name))
```

Comment:
@mmojzis I think we should check that the keypair also does not exist. So verify that we get the error message and check that the keypair was not added. What do you think?

Reply:
Well, that's not so easy. Keypair is not created immediately so I would have to create wait_for for at least a minute to check whether this keypair is not in keypair array. I don't think that makes much sense. This error response comes directly from provider so if keypair is created then it's problem on provider side and not in CFME.

Reply:
@mmojzis  Does CFME create a keypair record in CFME when a request is made? Does it delete when the request fails? I think it is worth validating that there is no records in vmdb and that nothing appears in CFME.

Reply:
No, that's why I have to wait until it's created. It calls provider and keypair is created on provider side. In CFME it's created by targeted refresh AFAIK. Provider sends data to CFME that it has a new record.
When request fails nothing is created as it's called directly from CFME UI to provider.

Reply:
Thanks @mmojzis

Reply:
@jawatts Rebased...But it started failing on one test for no reason....

Reply:
@mmojzis looks like it failed here:
```[KeyPairDetailsView/toolbar/configuration]: Selecting 'Download private key' (/usr/lib/python2.7/site-packages/widgetastic_patternfly/__init__.py:1684)```

So there is no "Download private key" button on the details page. See this screenshot:
http://10.16.45.124/aRzZLu1b/artifacts/cfme/tests/cloud/test_keypairs.py/test_download_private_key[openstack-13]/filedump-exception_screenshot.png

Reply:
@mmojzis looks like it is under the `Configuration` drop down

Reply:
@mmojzis It looks like it is because the "Download private key" selection is disabled. 
```
In[19]: view.toolbar.configuration.item_enabled('Download private key')
Out[20]: False
```

Reply:
@jawatts Thanks for helping me identifying that issue. I don't know how, but the problem was actually fixed by deleting some keypairs and it passed now.